pallets/pool-xyk/src/lib.rs

```diff
@@ -461,16 +469,30 @@ impl<T: Config> LiquiditySource<T::DEXId, T::AccountId, AssetIdOf<T>, Balance, D
         // Calculate quote.
         let (calculated, fee_amount) = match amount {
             QuoteAmount::WithDesiredInput { desired_amount_in } => {
-                Pallet::<T>::calc_output_for_exact_input(
+                let (calculated, fee_amount) = Pallet::<T>::calc_output_for_exact_input(
                     T::GetFee::get(),
                     get_fee_from_destination,
                     &reserve_input,
                     &reserve_output,
                     &desired_amount_in,
                     deduce_fee,
-                )?
+                )?;
+                let required_output = if get_fee_from_destination {
+                    calculated + fee_amount
+                } else {
+                    calculated
+                };
+                ensure!(
+                    required_output <= max_output_available,
+                    Error::<T>::NotEnoughOutputReserves
+                );
+                (calculated, fee_amount)
             }
             QuoteAmount::WithDesiredOutput { desired_amount_out } => {
+                ensure!(
+                    desired_amount_out <= max_output_available,
```

Comment:
In case when `max_output_available` == `reserve_output` == `desired_amount_out` it will fail when divided by 0 in `calc_input_for_exact_output()`.
Probably just `desired_amount_out < max_output_available` is better.

Also take a note that `calc_input_for_exact_output()` has a line:
```rust
let fxw_y_out = fxw_y_out.clone() + Fixed::from_bits(1);
```